components/workflow-mgt/org.wso2.carbon.identity.workflow.mgt/src/main/java/org/wso2/carbon/identity/workflow/mgt/WorkflowManagementServiceImpl.java

```diff
@@ -426,7 +426,56 @@ public void addAssociation(String associationName, String workflowId, String eve
         }
     }
 
+    /**
+     * List paginated Workflows with a filter
+     *
+     * @param tenantId Tenant Id
+     * @param limit Limit
+     * @param offset Offset
+     * @param filter Filter
+     * @return List<Workflow>
+     * @throws WorkflowException
+     */
+    @Override
+    public List<Workflow> listPaginatedWorkflows(int tenantId, int limit, int offset, String filter) throws WorkflowException{
+
+        List<WorkflowListener> workflowListenerList =
+                WorkflowServiceDataHolder.getInstance().getWorkflowListenerList();
+        for (WorkflowListener workflowListener : workflowListenerList) {
+            if (workflowListener.isEnable()) {
+                workflowListener.doPreListPaginatedWorkflows(tenantId, limit, offset, filter);
+            }
+        }
+        // Validate whether the limit is not zero or a negative number.
+        if (limit < 0) {
+            throw new WorkflowException(WFConstant.Exceptions.ERROR_INVALID_LIMIT);
```

Comment:
won't these be client-side Exceptions?

Don't we have a way to differentiate server and client exceptions? If not let's create sub exceptions from WorkflowException for the server side and client side.

Reply:
same for offset check

Reply:
Added WorkflowClientException class and handled client exceptions using that class.